crates/js-component-bindgen/src/core.rs

```diff
@@ -0,0 +1,794 @@
+//! Support for transpiling core modules using multi-memory to those that don't
+//! use multi-memory.
+//!
+//! Wasmtime's implementation of adapter modules between components requires the
+//! usage of multi-memory for copying data back and forth between two
+//! components. The multi-memory proposal is, at this time, not stable in any JS
+//! engine. This module is an attempt to polyfill this until at such a time that
+//! multi-memory can be used natively.
+//!
+//! The purpose of this module is to identify core wasms which require
+//! multi-memory coming out of Wasmtime. These wasms are rewritten to not
+//! actually use more than one memory. The implementation here is to replace all
+//! memory instructions operating on memory index 1 or greater with function
+//! calls where JS is the one that does the load/store/etc. This is not expected
+//! to be fast at runtime but is intended to be just enough to get this working
+//! in JS environments at this time. The true speed is expected to come with the
+//! multi-memory proposal.
+//!
+//! This module exports a [`Translation`] which wraps a [`ModuleTranslation`]
+//! either as a pass-through "normal" or an "augmented" version where
+//! "augmented" means that the original wasm is not used but instead a
+//! recompiled copy without multiple memories is used. When calculating the
+//! imports for the "augmented" module the arguments for JS functions that
+//! read/write memory are automatically injected and handled.
+//!
+//! Callers of this module need to have an implementation in JS for all of the
+//! entries listed in [`AugmentedOp`], likely through the `DataView` class in
+//! JS.
+//!
+//! Note that at this time this module is not intended to be a complete and
+//! general purpose method of compiling multiple memories to single-memory
+//! modules. This does not handle all instructions that use memory for example,
+//! but only those that Wasmtime's adapter modules emits. It's possible to add
+//! support for more instructions but such support isn't required at this time.
+//! Examples of unsupported instructions are `i64.load8_u` and `memory.copy`.
+//! Additionally core wasm sections such as data sections and tables are not
+//! supported because, again, Wasmtime doesn't use it at this time.
```

Comment:
I was wondering if it might make sense to use a full wasm2js solution here (eg something like https://github.com/evanw/polywasm), but this certainly seems adequate as well.

Reply:
I think any of those would work myself, this was just the first thing I thought of so I figured I'd take a stab implementing it while stuck delayed in an airport :)

If you'd prefer to go with those though I think it's ok to hold this open or close it, either way's fine by me